Text/CustomEntitySearch/CustomEntitySearch.cs

```diff
@@ -52,48 +53,273 @@ public static class CustomEntitySearch
+                            Confidence = 1
+                        });
+                }
+                addWord = true;
+            }
+
+            if (leniency > 0)
+            {
+                // using KMP for now
+                // Create Table
+                IList<int> KMPTable = (caseSensitive) ? CreateKMPTable(checkMatch) : CreateKMPTable(checkMatch.ToLower());
+
+                // Begin searching!
+                int currTextChar = 0;
+                int currWordChar = 0;
+                StringWriter wordFound = new StringWriter();
+                double currMismatch = 0;
+                IList<char> wordCharArray = (caseSensitive) ? checkMatch.ToCharArray() : checkMatch.ToLower().ToCharArray();
+                IList<char> textCharArray = (caseSensitive) ? text.ToCharArray() : text.ToLower().ToCharArray();
+                IList<string> valuesList = new List<string>();
+                IList<int> offsetList = new List<int>();
+                IList<double> confidenceList = new List<double>();
+                int minVal = -1;
+
+                while (currTextChar < textCharArray.Count)
+                {
+                    if (currWordChar == 0 && currMismatch > 0)
                     {
-                        words = new WordLinker(executionContext.FunctionAppDirectory).Words;
+                        wordFound.GetStringBuilder().Remove(0, 1);
+                        currMismatch--;
                     }
-                    
-                    var entities = new HashSet<Entity>();
-                    var entitiesFound = new HashSet<string>();
-                    if (!string.IsNullOrWhiteSpace(text))
+                    if (currMismatch <= leniency)
                     {
-                        foreach (string word in words)
+                        if (wordCharArray[currWordChar] == textCharArray[currTextChar])
                         {
-                            if (string.IsNullOrEmpty(word)) continue;
-                            string escapedWord = Regex.Escape(word);
-                            string pattern = @"\b(?ix:" + escapedWord + @")\b";
-                            MatchCollection entityMatch = Regex.Matches(text, pattern, RegexOptions.IgnoreCase, TimeSpan.FromSeconds(MaxRegexEvalTime));
-                            if (entityMatch.Count != 0)
+                            wordFound.Write(text.ElementAt(currTextChar));
+                            currTextChar++;
+                            currWordChar++;
+
+                            if (currWordChar >= wordCharArray.Count || (currTextChar >= textCharArray.Count 
+                                && currMismatch + (wordCharArray.Count - currWordChar) <= leniency))
                             {
-                                foreach (Match match in entityMatch)
+                                // Code Cleanup?
+                                if (currMismatch > 0 || entityMatch.Count == 0)
```

Comment:
>currMismatch > 0 [](start = 36, length = 16)

weird check, what does currMismatch < 0 represent?

Reply:
currMismatch = 0 is an exact match. sometimes, the regex matching doesn't find the exact match, so I had to add that case. currMismatch < 0 should not happen

Reply:
It looks like this is always true?

---
In reply to: [313017607](https://github.com/Azure-Samples/azure-search-power-skills/pull/5#discussion_r313017607) [](ancestors = 313017607)